library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -219,6 +288,117 @@ mkPrim (Types.Prim prim) ty = case prim of
+  ------------------------------
+  -- WARNING ∷ Leaking Memory
+  ------------------------------
+  env <- Closure.mallocEnvironment (fromIntegral (length xs))
+  let insertInto (Types.Capture {slot = Types.Slot {newIndex}, location}) = do
+        capturedArg <-
+          case location of
+            Types.FromAmbientEnv name -> do
+              var <- compileVar name
+              ------------------------------
+              -- WARNING ∷ Leaking Memory
+              ------------------------------
+              loc <- Block.mallocType (Typed.typeOf var)
+              Block.store loc var
+              pure loc
+            Types.FromClosureEnv index ->
+              loadElementIndex index
+        locPtr <-
+          Block.getElementPtr
+            Types.Minimal
+              { type' = Closure.environmentPtr,
+                address' = env,
+                indincies' =
+                  Block.constant32List [fromIntegral (Types.num newIndex)]
+              }
+        Block.store locPtr capturedArg
+  traverse_ insertInto xs
+  pure env
+
+--------------------------------------------------------------------------------
+-- Closure Argument Names
+--------------------------------------------------------------------------------
+
+closureArgumentName :: IsString p => p
+closureArgumentName = "juvix_argArray"
+
+closureCaptureName :: IsString p => p
+closureCaptureName = "juvix_environmentArray"
+
+--------------------------------------------------------------------------------
+-- Closure Indexing
+--------------------------------------------------------------------------------
+
+loadElementIndex :: Types.Call m => Types.IndexInto -> m LLVM.Operand
+loadElementIndex Types.IndexInto {index, into} = do
+  arr <-
+    case into of
+      Types.ClosureEnvironment ->
+        compileVar closureCaptureName
+      -- Does not happen in the current model, check in later for when
+      -- we can properly do this!
+      Types.ArgumentEnvironemnt ->
+        compileVar closureArgumentName
+  Block.loadElementPtr
+    Types.Minimal
+      { type' = Closure.environmentPtrDeref,
+        address' = arr,
+        indincies' = Block.constant32List [fromIntegral (Types.num index)]
+      }
+
+--------------------------------------------------------------------------------
+-- Function Definition Helpers
+--------------------------------------------------------------------------------
+
+compileFunctionEnv ::
+  Types.Define m =>
+  -- | Function name
+  Symbol ->
+  -- | The type of the lambda abstraction.
+  ErasedAnn.Type PrimTy ->
+  -- | List of parameter names.
+  [NameSymbol.T] ->
+  -- | The body of the lambda abstraction.
+  m a ->
+  m LLVM.Operand
+compileFunctionEnv name ty arguments body = do
+  let (llvmArgtyBeforeClosure, llvmRetty) =
+        functionTypeLLVM ty
+      llvmArgNames =
+        fmap (Block.internName . NameSymbol.toSymbol) arguments
+      -- turn the function pointers to closure for any HOF
```

Comment:
I'd probably spell out "higher-order functions" here.

Reply:
Feel free to make the commit